### DISCUSSION TITLE
Application dependent keybindings

### DISCUSSION THREAD

**Discussion by MartyMcFlyInTheSky** [wezterm/wezterm#6184] [Q&A]
### What Operating System(s) are you running on?

Linux X11

### Which Wayland compositor or X11 Window manager(s) are you using?

i3

### WezTerm version

wezterm 20240915-153243-2d0c5cdd

### Ask your question!

I want to forward certain keystrokes as escape sequences ONLY when nvim is running as foreground application behind the tty. Is this somehow possible? I know that there's the approach with 

```
    keys = {
        {
            key = 'K',
            mods = 'CTRL',
            action = wezterm.action_callback(function(window, pane)
                local process_name = pane:get_foreground_process_name()
                if process_name == 'nvim' then
                    window:perform_action(wezterm.action.SendKey({ key = 'K', mods = 'CTRL' }), pane)
                end
            end),
        },
        {
            key = 'J',
            mods = 'CTRL',
            action = wezterm.action_callback(function(window, pane)
                local process_name = pane:get_foreground_process_name()
                if process_name == 'nvim' then
                    window:perform_action(wezterm.action.SendKey({ key = 'J', mods = 'CTRL' }), pane)
                end
            end),
        },
    },
```

but this is so verbose and cumbersome if you have many keybindings! Is there  a better approach?

**Comment by wez** [comment]
not sure if it is better, but there are different tradeoffs if you use the update-status event to determine the foreground application, then use https://wezfurlong.org/wezterm/config/lua/window/set_config_overrides.html to override the keybindings

**Reply by MartyMcFlyInTheSky** [reply]
`set_config_overrides` looks tasty, but I don't like the polling nature of update-status. But makes me htink maybe there's a way to configure nvim to give of a startup/stop signal through the terminal so I can hook the config overrides to that signal :thinking:

**Comment by bew** [comment]
Note that you can always write a Lua function that returns a keybind, so you'd only have a single function call for each keybind you want to pass through.
👉 much less verbose and cumbersome

Something like: (not tested)
```lua
local function passthrough_keybind_for(process_name, key_info)
  return {
    key = key_info.key,
    mods = key_info.mods,
    action = wezterm.action_callback(function(window, pane)
      local fg_process_name = pane:get_foreground_process_name()
      if fg_process_name == process_name then
        window:perform_action(wezterm.action.SendKey(key_info), pane)
      end
    end),
  }
end

      -- ...
      keys = {
        passthrough_keybind_for("nvim", { key = "K", mods = "CTRL" }),
        -- ...
      }
```

**Reply by MartyMcFlyInTheSky** [reply]
True prob the easiest option, thx!

**Comment by LethargicDeveloper** [comment]
The issue I ran into using that approach was that keybinds were swallowed in other apps. So, for instance, if another app (other than nvim) had a CTRL+K keybind then nothing would actually happen because wezterm was just swallowing it. Here's an updated version that passing through keypresses to apps where you don't want to handle it. My setup below is a little different in that I want keybinds that only work in the shell prompt, otherwise, just pass it through. But you can update it for nvim.

There's got to be an easier way...

```
local function is_prompt(pane)
  local process_name = pane:get_foreground_process_name()
  return process_name == "/bin/zsh"
end

local function keybind(key, mods, func)
  return {
    key = key,
    mods = mods,
    action = wezterm.action_callback(function(window, pane)
      // only use custom keybind for specified program (shell prompt)
      if is_prompt(pane) then
        window:perform_action(func, pane)

      // otherwise, just pass the key press through so the app can handle it
      else
	window:perform_action(wezterm.action({ SendKey = { key = key, mods = mods } }), pane)
      end
    end),
  }
end

keys = {
  // load wezterm config in nvim
  keybind(
    ",",
    "CMD",
    act.SpawnCommandInNewTab({
      cwd = os.getenv("WEZTERM_CONFIG_DIR"),
      set_environment_variables = {
        TERM = "screen-256color",
      },
      args = {
        "/opt/homebrew/bin/nvim",
        os.getenv("WEZTERM_CONFIG_FILE"),
      },
    })
  ),
  keybind("Escape", "NONE", act.SendString("\x15")), // clear prompt
  keybind("Home", "NONE", act.SendString("\x01")),   // go to beginning of prompt
  keybind("End", "NONE", act.SendString("\x05")),    // go to end of prompt
},
```
